src/storage/remote/protocol.c

```diff
@@ -92,47 +93,120 @@ storageRemoteFilterGroup(IoFilterGroup *filterGroup, const Variant *filterList)
     if (info->level >= storageInfoLevelDetail)
     {
-        protocolServerWriteLine(server, jsonFromUInt(info->userId));
-        protocolServerWriteLine(server, jsonFromStr(info->user));
-        protocolServerWriteLine(server, jsonFromUInt(info->groupId));
-        protocolServerWriteLine(server, jsonFromStr(info->group));
-        protocolServerWriteLine(server, jsonFromUInt(info->mode));
+        // Write mode
+        pckWriteU32P(data->write, info->mode, .defaultNull = true, .defaultValue = data->modeLast);
+
+        // Write user id/name
+        pckWriteU32P(data->write, info->userId, .defaultNull = true, .defaultValue = data->userIdLast);
+
+        if (info->user == NULL)
+            pckWriteBoolP(data->write, true);
+        else
+        {
+            pckWriteNullP(data->write);
+            pckWriteStrP(data->write, info->user, .defaultNull = data->user != NULL, .defaultValue = data->user);
+        }
+
+        // Write group id/name
+        pckWriteU32P(data->write, info->groupId, .defaultNull = true, .defaultValue = data->groupIdLast);
 
+        if (info->group == NULL)
+            pckWriteBoolP(data->write, true);
+        else
+        {
+            pckWriteNullP(data->write);
+            pckWriteStrP(data->write, info->group, .defaultNull = data->group != NULL, .defaultValue = data->group);
+        }
+
+        // Write link destination
         if (info->type == storageTypeLink)
-            protocolServerWriteLine(server, jsonFromStr(info->linkDestination));
+            pckWriteStrP(data->write, info->linkDestination);
+    }
+
+    // Store defaults to use for the next call. If memContext is NULL this function is only being called one time so there is no
+    // point in storing defaults.
+    if (data->memContext != NULL)
+    {
+        data->timeModifiedLast = info->timeModified;
+        data->modeLast = info->mode;
+        data->userIdLast = info->userId;
+        data->groupIdLast = info->groupId;
+
+        if (!strEq(info->user, data->user) && info->user != NULL)
+        {
+            strFree(data->user);
+
+            MEM_CONTEXT_BEGIN(data->memContext)
+            {
+                data->user = strDup(info->user);
+            }
+            MEM_CONTEXT_END();
+        }
+
+        if (!strEq(info->group, data->group) && info->group != NULL)
+        {
+            strFree(data->group);
+
+            MEM_CONTEXT_BEGIN(data->memContext)
+            {
+                data->group = strDup(info->group);
+            }
+            MEM_CONTEXT_END();
+        }
     }
 
     FUNCTION_TEST_RETURN_VOID();
 }
 
 static void
-storageRemoteProtocolInfoListCallback(void *server, const StorageInfo *info)
+storageRemoteProtocolInfoListCallback(void *dataVoid, const StorageInfo *info)
 {
     FUNCTION_TEST_BEGIN();
-        FUNCTION_LOG_PARAM(PROTOCOL_SERVER, server);
+        FUNCTION_LOG_PARAM_P(VOID, dataVoid);
         FUNCTION_LOG_PARAM(STORAGE_INFO, info);
     FUNCTION_TEST_END();
```

Comment:
Same comment:
1) Can *data or *info be NULL? If not, maybe need ASSERTs?
2) Why is FUNCTION_TEST_PARAM(STORAGE_INFO, info); not with an _P:  FUNCTION_TEST_PARAM_P(STORAGE_INFO, info);

Reply:
1. Asserts added.
2. Because STORAGE_INFO is a defined logging type (see storage/info.h) that happens to be a pointer but we don't need to know that here. void * is not a define logging type so we need to specify _P to let the macro know it is a pointer.